### ISSUE TITLE
Issue in the creation of the design matrices for the random walk

### ISSUE THREAD

**Issue by seabbs** [epinowcast/epinowcast#166]
I wonder if there is an issue in the creation of the design matrices for the random walk in this PR. For the f-matrix we add a "c" to the variable name, ie if we have rw(week) we use cweekX as column names, but we do not add this "c" in the r-matrix. This means that the variable order can be different in the two matrices and we don't use the correct variances for the right betas. In my attempt at making a simple forecasting implementation, I think it worked to just remove the "c" prefix in the two functions used to generate the random walk design matrix. Possible that I have overlooked something important and that this is not an problem or that another solution is better.

_Originally posted by @Gulfa in https://github.com/epinowcast/epinowcast/issues/152#issuecomment-1235390723_

**Comment by seabbs** [comment]
Thanks for flagging this @gulfa. 

I've promoted this to a general issue as it would impact all random walk use cases not just those in #152. 

When implementing this feature I think this is accounted for in the reliance on ordering of the matrices (so direct name matches aren't needed). I could very well be wrong here. Do you see differences in behaviour with the two formulations?

**Comment by seabbs** [comment]
Thanks, @Gulfa. 

It does sound like you may be correct. I will explore a few toy examples to check and circle back.

**Comment by seabbs** [comment]
Hi @Gulfa, 

Sorry for the big delay in getting to this. I've dug into this a little bit and would appreciate hearing your thoughts.

So I am splitting this into two points:

1.  For the f-matrix we add a "c" to the variable name, ie if we have rw(week) we use cweekX as column names, but we do not add this "c" in the r-matrix. This means that the variable order can be different in the two matrices and we don't use the correct variances for the right betas.

*So I agree this is indeed the case but I am not sure I follow the "this means". The choice to not add the `c` was to maintain the variable names that users put in (for future improved postprocessing etc) but obviously, this prevents the use of a model like `~ (1 | week) + rw(week)` (though this currently errors for an unrelated reason). That trade-off seemed okay at the time as the models it rules out are quite fringe and can still be implemented using a variable renaming.*

*Digging into the code a bit I see that the variable name is used as a label for the random walk random effects but the new variable (i.e `cvariable`) is used when actually allocating random effects (https://github.com/epinowcast/epinowcast/blob/c36d397e67ef17deb8843008b5b2d44298f8f5ae/R/formula-tools.R#L342).*

*For the overall construction of the random effects design matrix (i.e this quite convoluted code https://github.com/epinowcast/epinowcast/blob/c36d397e67ef17deb8843008b5b2d44298f8f5ae/R/formula-tools.R#L700) the non-modified variable name is used (as you say) but as this is what is used when constructing the random walk random effects this all works as expected?*

2. I think it worked to just remove the "c" prefix in the two functions used to generate the random walk design matrix.

*So we likely need some kind of identifier for these created variables as otherwise formulas that combine multiple versions of the same variable in different ways (i.e in the below example) could have issues.*

Here is a small example showing a simple random walk combined with a fixed effect for week. I think the potential bug you flagged would cause an issue here but from what I can see it looks okay? We may just need a different reprex to get to the bottom of this so happy to take pointers.

```r
# Use meta data for references dates from the Germany COVID-19
# hospitalisation data.
obs <- enw_filter_report_dates(
  germany_covid19_hosp[location == "DE"],
  remove_days = 20
)
obs <- enw_filter_reference_dates(obs, include_days = 20)
pobs <- enw_preprocess_data(obs, by = c("age_group", "location"), max_delay = 5)
data <- pobs$metareference[[1]]

# Model with a random effect for age group and a random walk
enw_formula(~ week + rw(week), data)
```
```
$formula
[1] "~week + rw(week)"

$parsed_formula
$parsed_formula$fixed
[1] "week"

$parsed_formula$random
NULL

$parsed_formula$rw
[1] "rw(week)"


$expanded_formula
[1] "~week + cweek1 + cweek2"

$fixed
$fixed$formula
[1] "~week + cweek1 + cweek2"

$fixed$design
   (Intercept) week cweek1 cweek2
1            1    0      0      0
8            1    1      1      0
15           1    2      1      1

$fixed$index
  [1] 1 1 1 1 1 1 1 2 2 2 2 2 2 2 3 3 3 3 3 3 3 1 1 1 1 1 1 1 2 2 2 2 2 2 2 3 3 3 3 3 3 3 1 1
 [45] 1 1 1 1 1 2 2 2 2 2 2 2 3 3 3 3 3 3 3 1 1 1 1 1 1 1 2 2 2 2 2 2 2 3 3 3 3 3 3 3 1 1 1 1
 [89] 1 1 1 2 2 2 2 2 2 2 3 3 3 3 3 3 3 1 1 1 1 1 1 1 2 2 2 2 2 2 2 3 3 3 3 3 3 3 1 1 1 1 1 1
[133] 1 2 2 2 2 2 2 2 3 3 3 3 3 3 3


$random
$random$formula
[1] "~0 + fixed + week"

$random$design
  fixed week
1     1    0
2     0    1
3     0    1
attr(,"assign")
[1] 1 2

$random$index
[1] 1 2 3


attr(,"class")
[1] "enw_formula" "list" 
```

Diving in a little deeper and just looking at `construct_rw()`.  We can see that the random effects `data.frame` is using `week` not `cweek` (which I agree might be confusing and so could arguably be changed).

```r
epinowcast:::construct_rw(rw(week), data)
```
```
$data
Key: <.group, date>
           date .group age_group location delay day_of_week   day  week month cweek1 cweek2
         <IDat>  <int>    <fctr>   <fctr> <num>      <fctr> <num> <num> <num>  <num>  <num>
  1: 2021-09-10      1       00+       DE     0      Friday     0     0     0      0      0
  2: 2021-09-11      1       00+       DE     0    Saturday     1     0     0      0      0
  3: 2021-09-12      1       00+       DE     0      Sunday     2     0     0      0      0
  4: 2021-09-13      1       00+       DE     0      Monday     3     0     0      0      0
  5: 2021-09-14      1       00+       DE     0     Tuesday     4     0     0      0      0
 ---                                                                                       
143: 2021-09-26      7       80+       DE     0      Sunday    16     2     0      1      1
144: 2021-09-27      7       80+       DE     0      Monday    17     2     0      1      1
145: 2021-09-28      7       80+       DE     0     Tuesday    18     2     0      1      1
146: 2021-09-29      7       80+       DE     0   Wednesday    19     2     0      1      1
147: 2021-09-30      7       80+       DE     0    Thursday    20     2     0      1      1

$terms
[1] "cweek1" "cweek2"

$effects
   effects fixed  week
    <char> <num> <num>
1:  cweek1     0     1
2:  cweek2     0     1
```

**Comment by seabbs** [comment]
I haven't made much progress on my understanding of this. @Gulfa (or anyone else) any light here as it would be great to fix any bugs or if we are 🐛  free (at least here) allow me to save my worry for other pressing issues.... 🙀

**Comment by seabbs** [comment]
I think I have understood this now and there is a bug when there is both a random effect and a random walk being used as below. 

This is because the random effects are added to the fixed effects design matrix as the last action whilst random effects are added to the random effects metadata second. 

``` r
library(epinowcast)
# U/se meta data for references dates from the Germany COVID-19
# hospitalisation data.
obs <- enw_filter_report_dates(
  germany_covid19_hosp[location == "DE"],
  remove_days = 40
)
obs <- enw_filter_reference_dates(obs, include_days = 40)
pobs <- enw_preprocess_data(obs, by = c("age_group", "location"))
data <- pobs$metareference[[1]]

# Model with a random effect for age group and a random walk
enw_formula(~ 1 + (1 | age_group) + rw(week), data)
#> $formula
#> [1] "~1 + (1 | age_group) + rw(week)"
#>
#> $parsed_formula
#> $parsed_formula$fixed
#> [1] "1"
#>
#> $parsed_formula$random
#> $parsed_formula$random[[1]]
#> 1 | age_group
#>
#>
#> $parsed_formula$rw
#> [1] "rw(week)"
#>
#>
#> $expanded_formula
#> [1] "~1 + cweek1 + cweek2 + cweek3 + cweek4 + cweek5 + age_group"
#>
#> $fixed
#> $fixed$formula
#> [1] "~1 + cweek1 + cweek2 + cweek3 + cweek4 + cweek5 + age_group"
#>
#> $fixed$design
#>     (Intercept) cweek1 cweek2 cweek3 cweek4 cweek5 age_group00-04 age_group00+
#> 1             1      0      0      0      0      0              0            1
#> 8             1      1      0      0      0      0              0            1
#> 15            1      1      1      0      0      0              0            1
#> 22            1      1      1      1      0      0              0            1
#> 29            1      1      1      1      1      0              0            1
#> 36            1      1      1      1      1      1              0            1
#> 42            1      0      0      0      0      0              1            0
#> 49            1      1      0      0      0      0              1            0
#> 56            1      1      1      0      0      0              1            0
#> 63            1      1      1      1      0      0              1            0
#> 70            1      1      1      1      1      0              1            0
#> 77            1      1      1      1      1      1              1            0
#> 83            1      0      0      0      0      0              0            0
#> 90            1      1      0      0      0      0              0            0
#> 97            1      1      1      0      0      0              0            0
#> 104           1      1      1      1      0      0              0            0
#> 111           1      1      1      1      1      0              0            0
#> 118           1      1      1      1      1      1              0            0
#> 124           1      0      0      0      0      0              0            0
#> 131           1      1      0      0      0      0              0            0
#> 138           1      1      1      0      0      0              0            0
#> 145           1      1      1      1      0      0              0            0
#> 152           1      1      1      1      1      0              0            0
#> 159           1      1      1      1      1      1              0            0
#> 165           1      0      0      0      0      0              0            0
#> 172           1      1      0      0      0      0              0            0
#> 179           1      1      1      0      0      0              0            0
#> 186           1      1      1      1      0      0              0            0
#> 193           1      1      1      1      1      0              0            0
#> 200           1      1      1      1      1      1              0            0
#> 206           1      0      0      0      0      0              0            0
#> 213           1      1      0      0      0      0              0            0
#> 220           1      1      1      0      0      0              0            0
#> 227           1      1      1      1      0      0              0            0
#> 234              0              0              0              1            0
#> 241              0              0              0              1            0
#> 247              0              0              0              0            1
#> 254              0              0              0              0            1
#> 261              0              0              0              0            1
#> 268              0              0              0              0            1
#> 275              0              0              0              0            1
#> 282              0              0              0              0            1
#>
#> $fixed$index
#>   [1]  1  1  1  1  1  1  1  2  2  2  2  2  2  2  3  3  3  3  3  3  3  4  4  4  4
#>  [26]  4  4  4  5  5  5  5  5  5  5  6  6  6  6  6  6  7  7  7  7  7  7  7  8  8
#>  [51]  8  8  8  8  8  9  9  9  9  9  9  9 10 10 10 10 10 10 10 11 11 11 11 11 11
#>  [76] 11 12 12 12 12 12 12 13 13 13 13 13 13 13 14 14 14 14 14 14 14 15 15 15 15
#> [101] 15 15 15 16 16 16 16 16 16 16 17 17 17 17 17 17 17 18 18 18 18 18 18 19 19
#> [126] 19 19 19 19 19 20 20 20 20 20 20 20 21 21 21 21 21 21 21 22 22 22 22 22 22
#> [151] 22 23 23 23 23 23 23 23 24 24 24 24 24 24 25 25 25 25 25 25 25 26 26 26 26
#> [176] 26 26 26 27 27 27 27 27 27 27 28 28 28 28 28 28 28 29 29 29 29 29 29 29 30
#> [201] 30 30 30 30 30 31 31 31 31 31 31 31 32 32 32 32 32 32 32 33 33 33 33 33 33
#> [226] 33 34 34 34 34 34 34 34 35 35 35 35 35 35 35 36 36 36 36 36 36 37 37 37 37
#> [251] 37 37 37 38 38 38 38 38 38 38 39 39 39 39 39 39 39 40 40 40 40 40 40 40 41
#> [276] 41 41 41 41 41 41 42 42 42 42 42 42
#>
#>
#> $random
#> $random$formula
#> [1] "~0 + fixed + age_group + week"
#>
#> $random$design
#>    fixed age_group week
#> 1      0         1    0
#> 2      0         1    0
#> 3      0         1    0
#> 4      0         1    0
#> 5      0         1    0
#> 6      0         1    0
#> 7      0         1    0
#> 8      0         0    1
#> 9      0         0    1
#> 10     0         0    1
#> 11     0         0    1
#> 12     0         0    1
#> attr(,"assign")
#> [1] 1 2 3
#>
#> $random$index
#>  [1]  1  2  3  4  5  6  7  8  9 10 11 12
#>
#>
#> attr(,"class")
#> [1] "enw_formula" "list"
```

<sup>Created on 2023-04-11 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by seabbs** [comment]
@Gulfa I think I resolved this in #228. Would be great to know if you agree!
